[WIP] Remove documented algebra

Krever opened the pull request:
This is ongoing experiment to see if we can, in fact, unify documented and undocumented algebras. The main motivation for these were some comments from people trying to use endpoints as well as my own experience when preparing for scalar.

Unfortunately, this will not be a very atomic changeset as it contains a couple of different things that I added mostly because I could:

- xmap as described in #56.
- experimental mill build. It is much more convenient for development, especially when I can easily go one module after another. Importing sbt build to intellij takes more than 10 minutes and importing mill takes less than a minute.
- ability to combine request headers

To Do:
- [x] go through all methods and verify defaults for documentation
- [x] cover newly added requests with tests(`xmap`s, headers algebra, headers semigroup).
- [x] adjust rest of interpreters
- [x] Fix json entites tests
- [x] Review TODOs
- [x] Fix sbt build
- [x] Update documentation & examples

julienrf commented:
Interesting!

One thing that bothers me with sbt is that it always tries to re-resolve dependencies when you compile and it seems to take forever. Does that work better with mill?

Regarding your changes, do you plan to keep empty strings as default values for documentation? I think that would be really confusing.

Krever commented:
Mill has put a lot of focus on caching and it basically caches every task, and dependencies is a task as well, so it gets resolved only when it changes.

I plan to have options for everything that is optional in openapi, and probably empty strings for required fields, if there are any

Krever commented:
I reviewed, hopefully, all of the documentation parameters and there are 3 things we need to decide:

- do we want to have some more specific type for docs, like `type Docs = Option[String] with Xxx` or just a type alias? Do we want to define some string interpolator that will construct `Option[String]` so the user doesn't have to type `Some` everywhere?
- should we use a common name for this kind of parameters? Candidates are `documentation`, `description`, `doc`, `desc`. I would prefer something shorter but `doc` may be misleading as it is a common shortcut for `document`. That's why I vote for `desc` and it also plays well with openapi where there are multiple docs string(e.g. operation object has both `summary` and `description`).
- should we use string for non-optional documentation and `""` as a default? The only place where the description is not optional in openapi is response description. An alternative would be to hide this as "implementation detail" and expose uniform `Option[String]`, but this may be confusing and I would rather vote for 1:1 reflection openapi API but I'm not strong on this.

I also added vocabulary to describe headers

julienrf commented:
> do we want to have some more specific type for docs, like type Docs = Option[String] with Xxx or just a type alias? Do we want to define some string interpolator that will construct Option[String] so the user doesn't have to type Some everywhere?

Oh the string interpolator idea is nice, why not. It seems that we only need a `String` or an `Option[String]` when the documentation is optional, what would we get by having a `with Xxx`?

> should we use a common name for this kind of parameters? Candidates are documentation, description, doc, desc. I would prefer something shorter but doc may be misleading as it is a common shortcut for document. That's why I vote for desc and it also plays well with openapi where there are multiple docs string(e.g. operation object has both summary and description).

My reasoning was that we already use the word “description” for the code that uses the algebra interfaces : they describe communication endpoints, and these description can have several interpretations. Therefore, “documentation” seems to be the most appropriate, because, as you mentioned, “doc” could mean “document” as well. But, it’s true that it’s long :(

> should we use string for non-optional documentation and "" as a default? The only place where the description is not optional in openapi is response description. An alternative would be to hide this as "implementation detail" and expose uniform Option[String], but this may be confusing and I would rather vote for 1:1 reflection openapi API but I'm not strong on this.

I really dislike having empty strings as default values. I think this is like programming with `null`: we will have to carefully check whether a description is empty or not, at use-site, and chances are that we will forget to do so. I’d rather go with an `Option` with a `None` default value, or no default value at all.

Krever commented:
> About mill: I’m happy to drop the sbt build if we can support everything in Mill.
Happy to hear, but it will take some time probably. For now, I would like to experiment with it a little more and see how well it works. I don't feel like its ready to replace sbt but its definitely faster in build-test loop. In the end, I want to have this PR built with sbt.

Fristi commented:
Awesome work @Krever ! 

I would suggest to keep the changes per MR minimal :-) Small patches will not stale in the merge request stage very long due reviews or on going work.

Fristi commented:
Cool, I'm curious how it will look like 👍

Krever commented:
Few new changes:
* For now I decided to go with `docs: Documentation` as a general patter for docs, `type Documentation = Option[String]`. We can rename later if there is any better fit.
* For documentation required by openapi our api is uniform, so we still take `Option[String]` and apply default later. Its mostly because a lot of users are not gonna generate docs.
* I removed `testsuite` project and moved all elements to appropriate projects tests
* I added tests for newly introduced elements
* I introduced `textRequest` to define `RequestEntity[String]`
* *Mill build is finally usable*. I documented how to get the most important pieces, intellij support, compile and test. It's incredibly fast, at least now with the subset of projects.

Now I will probably start going through other interpreters one-by-one.

julienrf commented:
> we still take Option[String] and apply default later

Which default? And how to prevent users to forget about supplying a required documentation?

This is my main concern about that topic.

Krever commented:
> Which default? And how to prevent users from forgetting about supplying a required documentation?

The default will be an empty string, but let's analyze this a bit. From my perspective, we have two main users:

- someone who doesn't need any documentation - for him the default doesn't really matter as it won't ever be used. API needs to be as simple as possible and should allow to start using documentation seamlessly as the requirement for it may pop up unexpectedly.
- someone who deliberately used the openapi interpreter and want to have good documentation 

When defining endpoints the second user will face the following dangers:

- accidentally omitting an optional element of the documentation - with the current strategy it will be omitted when generating JSON
- accidentally omitting required element - with the current strategy it will be rendered with an empty string

So, in my opinion, these two cases are equal. In both, we generate a valid document with a little bit less information that will still work with any tools consuming this document. We are not able to predict which information user wanted to have and forgot and which of them skipped intentionally.

We could try to do something more sophisticated here, like customizing defaults, or lack of them, with implicits or fail to generate docs at runtime when it is not provided or print some warning. I think neither of this approaches is worth the effort as I can easily imagine someone who just doesn't want to specify the description of some of the responses. I don't know why openapi spec says its required, but from the perspective of our users, it is not.

Krever commented:
I have gone through openapi spec and I found only 2 fields that are required and that are somehow touched by this PR:
* response description
* path parameter name

For response description, I believe we can just leave it empty. No harm done done.
For path parameter(segment) name: I implemented autogeneration of this name. So segments with empty names will be substituted with `_arg1`, `_arg2` and so on. We can discuss the details but in general, I believe this is a good approach.

julienrf commented:
You can re-split some modules if that makes more sense. I remember that I’ve put together the algebra and the OpenAPI interpreter in the same module, for some modules, just because it was simpler.

Krever commented on documentation/examples/documented/src/main/scala/counter/Counter.scala:
```diff
@@ -33,7 +34,7 @@ import endpoints.documented.{algebra, generic}
   extends algebra.Endpoints
-    with algebra.JsonSchemaEntities
+    with JsonSchemaEntities
     with generic.JsonSchemas {
```
I have no idea how this changed landed as I have not yet touched(knowingly) `documentation` folder.

Krever commented on algebras/algebra/src/main/scala/endpoints/SemigroupK.scala:
```diff
@@ -0,0 +1,15 @@
+
+import scala.language.higherKinds
+
+trait SemigroupK[F[_]] {
```
Thanks, I could have checked. My unfamiliarity with cats shows up :) Will rename to `Semigroupal`

julienrf commented:
Haha, this probably a joke from Intellij then. It does that kind of things sometimes. 

On April 26, 2018 8:39:16 PM GMT+02:00, "Wojtek Pituła" <notifications@github.com> wrote:
>Krever commented on this pull request.
>
>
>
>> @@ -33,7 +34,7 @@ import endpoints.documented.{algebra, generic}
> 
> trait CounterEndpoints
>   extends algebra.Endpoints
>-    with algebra.JsonSchemaEntities
>+    with JsonSchemaEntities
>     with generic.JsonSchemas {
>
>I have no idea how this changed landed as I have not yet
>touched(knowingly) `documentation` folder.

Krever commented on play/server/src/main/scala/endpoints/play/server/Endpoints.scala:
```diff
@@ -168,14 +201,24 @@ trait Endpoints extends algebra.Endpoints with Urls with Methods {
   /** A successful HTTP response (status code 200) with no entity */
-  lazy val emptyResponse: Response[Unit] = _ => Results.Ok
+  def emptyResponse(docs: Documentation): Response[Unit] = _emptyResponse
+  lazy val _emptyResponse: Response[Unit] = _ => Results.Ok
```
fixed in, hopefully, all palces

Krever commented on openapi/openapi/src/test/scala/endpoints/openapi/EndpointsTest.scala:
```diff
@@ -39,15 +39,39 @@ class EndpointsTest extends WordSpec with Matchers with OptionValues {
+    "be substituted with auto generated name" in {
+      val path = Fixtures.documentation.paths.find(_._1.startsWith("/emptySegmentNameEndp")).get
+
+      path._1 shouldEqual "/emptySegmentNameEndp/{_arg2}"
```
Yes,  done.

Krever commented:
Ok, so it works now. I updated the documentation a bit, so it no longer mentions delegation. Bot, code and docs, could probably be polished further but I think this is PR is big enough right now. 

@julienrf Do you think we can merge this?

julienrf commented on documentation/manual/src/ornate/tutorial-openapi.md:
```diff
@@ -158,13 +156,9 @@ Finally, we start a `NettyServer` and give it both the `CounterServer` and
-must be written using a different set of algebra interfaces, which live
-in the `endpoints.documented.algebra` package.
+must be written using the same set of algebra interfaces, but with additional
+parameters supplied.
```
I wouldn’t mention that we use the same set of algebra interfaces. Only the fact that generating the documentation just consists in applying another interpreter to our endpoint descriptions, and that we can enrich those descriptions with additional documentation information.

julienrf commented on documentation/manual/src/ornate/tutorial.md:
```diff
@@ -580,7 +580,7 @@ and since we are only describing the endpoint, we added an abstract implicit met
 `option(jsonResponse[Meter])`, which means that it
 can optionally be empty (for instance if a client queries this endpoint with a
-non-existing `UUID`). The `option` method, provided by the `OptionalResponses` algebra
+non-existing `UUID`). The `option` method, provided by the `Responses` algebra
```
I would just drop the “, provided by the `Responses` algebra interface, ” because now it is part of the basic `Endpoints` algebra interface.

Krever replied:
done

julienrf commented on documentation/manual/src/ornate/tutorial.md:
```diff
@@ -629,7 +629,7 @@ In this section we have seen that deriving a Scala.js client works the same way
 
 We have seen that the endpoints library provides a minimal infrastructure that is designed to
-be extended according to application-specific needs. In our case we saw how the `OptionalResponses`
+be extended according to application-specific needs. In our case we saw how the `OptionalResponses` //TODO OptionalResponses are no longer there
```
Oops… we have to find another example… Maybe `Assets` or `AuthenticatedEndpoints`?

julienrf commented on openapi/build.sbt:
```diff
@@ -1,61 +1,52 @@
-val `json-schema-jvm` = `json-schema`.jvm
-
-val `json-schema-generic` =
+lazy val `json-schema-generic` =
```
I think this should be moved to the `algebra` source tree.

Krever replied:
Thats a good point. I think they should but on the other hand the openapi falvour of json schema is not standard and thats why I left it here. Cant say if someone wont be confused by the differences if we dont explicitly connect this implementation with `openapi`. What do you think?

julienrf replied:
OK then.

julienrf commented on openapi/build.sbt:
```diff
@@ -1,61 +1,52 @@
   crossProject.crossType(CrossType.Pure).in(file("json-schema-circe"))
     .settings(publishSettings ++ `scala 2.10 to 2.12`: _*)
     .settings(
       name := "endpoints-openapi-json-schema-circe",
```
This one is not coupled with OpenAPI, is it? I think we should find another name for it.

Krever replied:
renamed and moved

julienrf commented on openapi/openapi/src/main/scala/endpoints/openapi/Endpoints.scala:
```diff
@@ -0,0 +1,87 @@
+      val prefix = "_arg"
+      request.url.path
+        .foldLeft((0, List[Either[String, DocumentedParameter]]())) { // (num of empty-named args, new args list)
+          case ((nextEmptyArgNum, args), Right(arg)) if arg.name.isEmpty =>
```
In case you like the magic value suggestion, just replace `arg.name.isEmpty` with `arg.name == emptySegmentName`.

julienrf commented on openapi/openapi/src/main/scala/endpoints/openapi/Endpoints.scala:
```diff
@@ -0,0 +1,87 @@
+        response.map(r => r.status -> Response(r.documentation, r.content)).toMap
+      )
+    val item = PathItem(Map(method -> operation))
+    val path = correctPathSegments.collect {
```
Do you mind using `map` since the transformation function is total?

Krever replied:
done

julienrf commented on openapi/openapi/src/main/scala/endpoints/openapi/Urls.scala:
```diff
@@ -18,56 +20,59 @@ trait Urls extends algebra.Urls {
+    DocumentedQueryString(List(DocumentedParameter(name, required = false, docs, value)))
 
-  type QueryStringParam[A] = Unit
+  type QueryStringParam[A] = Schema
```
:+1:

julienrf commented on play/server/src/main/scala/endpoints/play/server/Endpoints.scala:
```diff
@@ -168,14 +201,24 @@ trait Endpoints extends algebra.Endpoints with Urls with Methods {
   /** A successful HTTP response (status code 200) with string entity */
-  lazy val textResponse: Response[String] = x => Results.Ok(x)
+  def textResponse(docs: Documentation): Response[String] = _textResponse
+  private lazy val _textResponse: Response[String] = x => Results.Ok(x)
```
Here again I would just inline `_emptyResponse` and `_textResponse`.

Krever replied:
done

julienrf commented:
@Krever I’ve left several comments, some of them should be addressed in this PR but we can defer some others to further PRs. In such a case can you create an issue to keep track of them?

cb372 commented:
@Krever I just want to give you an encouraging cheer and say, I hope you can get this over the line soon! It looks really good.

Krever commented on akka-http/client/src/main/scala/endpoints/akkahttp/client/Endpoints.scala:
```diff
@@ -40,14 +91,16 @@ class Endpoints(val settings: EndpointsSettings)
   type Response[A] = HttpResponse => Future[Either[Throwable, A]]
 
-  val emptyResponse: Response[Unit] = x =>
+  private lazy val _emptyResponse: Response[Unit] = x =>
```
Sure, I originally did that for performance reason but I will be happy to simplify. Do you think we shouldnt bother?

Krever replied:
done

Krever commented on algebras/algebra/src/main/scala/endpoints/algebra/Methods.scala:
```diff
@@ -14,4 +14,6 @@ trait Methods {
 
   def Patch: Method
 
+  def Options: Method
```
I dont follow. I added the `options` http method because sooner or later someone will expect to have it. It has nothing to do with removing the algebra but its better to have it than dont, isnt it?

julienrf replied:
Sorry, my comment was a bit short :)

The first part was: is it needed in order to remove the documented algebra? But I guess not :)

The second part was a thought: I think our server interpreters could provide `options` requests handling based on the descriptions of the endpoints. The idea would be that users wouldn’t have to explicitly describe endpoints with the `options` method, but servers would be able to reply to such requests in a useful way.

Krever commented on algebras/algebra/src/main/scala/endpoints/algebra/Urls.scala:
```diff
@@ -97,6 +97,9 @@ trait Urls {
   type Path[A] <: Url[A]
 
+  /** Implicit conversion to get rid of intellij errors when defining paths. Effectively should not be called.*/
+  implicit def dummyPathToUrl[A](p: Path[A]): Url[A] = p
```
So I kind of agree in principle, but I've seen at least 3 times people thinking they have something wrong when being introduced to endpoints because intellij marks it as red. It doesnt make the experience better.

julienrf replied:
OK then!

Krever commented on algebras/algebra/src/main/scala/endpoints/algebra/Urls.scala:
```diff
@@ -108,10 +111,10 @@ trait Urls {
 
   /** Builds a path segment carrying an `A` information */
-  def segment[A](implicit s: Segment[A]): Path[A]
+  def segment[A](name: String = "", docs: Documentation = None)(implicit s: Segment[A]): Path[A]
```
I would rather have a single way to do things, to not introduce too much confusion to the user. Do you think there is a lot of value in it?

Reg magic value, sure but, what is the benefit over empty string?

julienrf replied:
> I would rather have a single way to do things, to not introduce too much confusion to the user. Do you think there is a lot of value in it?

No, you’re right, let’s start with what you did!

> Reg magic value, sure but, what is the benefit over empty string?

It’s probably highly subjective so it’s ok if you don’t agree with me: I tend to exercise caution with using an empty string to represent some meaningful information. I feel like it’s like using `null` to represent the absence of value, it might make it harder to understand or find issues in our code.

Krever replied:
I would go with special value if empty string was acceptable value for open api, but I think it isn't, so it makes sense to handle it specially, IMHO.

Krever commented on documentation/examples/documented/src/main/scala/counter/Counter.scala:
```diff
@@ -54,54 +56,64 @@ trait CounterEndpoints
   // describes that the case class `Counter` has one field
   // of type `Int` named “value”
-  implicit lazy val jsonSchemaCounter: JsonSchema[Counter] = genericJsonSchema
+  implicit def jsonSchemaCounter: JsonSchema[Counter]
```
Do you have an idea how to fix it? We cannot reuse the schema interpretation because in openapi its real schema and in server its circe codec generated via schema api. To implement desired behaviour we would need an ability to derive circe codec from already materialized openapi schema.

julienrf replied:
> To implement desired behaviour we would need an ability to derive circe codec from already materialized openapi schema.

Yes, we used to have such an interpreter: https://static.javadoc.io/org.julienrf/endpoints-openapi-json-schema-circe_2.12/0.4.0/endpoints/documented/circe/JsonSchemas.html

Krever replied:
Wow, must have missed it, certainly need to fix it.

Krever commented on documentation/examples/overview/endpoints/src/main/scala/overview/CounterEndpoints.scala:
```diff
@@ -18,15 +18,15 @@ trait CounterEndpoints extends Endpoints with JsonEntitiesFromCodec {
     * The response entity is a JSON document representing the counter value.
     */
-  val currentValue = endpoint(get(path / "current-value"), jsonResponse[Counter])
+  val currentValue = endpoint(get(path / "current-value"), jsonResponse[Counter]())
```
I'm willing to add them if you think they really bring some value. In other case I wouldnt do that in sake of simplicity.

Krever commented on akka-http/server/src/main/scala/endpoints/akkahttp/server/Endpoints.scala:
```diff
@@ -35,13 +37,47 @@ trait Endpoints extends algebra.Endpoints with Urls with Methods {
+      RESPONSES
+  ************************* */
+
+  def emptyResponse(docs: Documentation): Response[Unit] = x => Directives.complete((StatusCodes.OK, ""))
```
Good catch, fixed

Krever commented on algebras/build.sbt:
```diff
@@ -1,11 +1,28 @@
+  crossProject.crossType(CrossType.Pure).in(file("json-schema"))
+    .settings(publishSettings ++ `scala 2.10 to 2.12`: _*)
+    .settings(
+      name := "endpoints-openapi-json-schema",
```
renamed

Krever commented on algebras/build.sbt:
```diff
@@ -1,11 +1,28 @@
+      libraryDependencies ++= Seq(
+        "com.github.tomakehurst" % "wiremock" % "2.6.0" % Test,
+        "org.scalatest" %%% "scalatest" % scalaTestVersion % Test,
+        compilerPlugin("org.scalamacros" % "paradise" % "2.1.0" % Test cross CrossVersion.full)
```
nope, removed

Krever commented on documentation/manual/src/ornate/interfaces-and-interpreters.md:
```diff
@@ -28,18 +28,9 @@ navigate through the algebra interfaces hierarchy.
-by the [endpoints-openapi](https://index.scala-lang.org/julienrf/endpoints/endpoints-openapi)
-artifact. This artifact also provides its interpreters (in the
-[endpoints.documented.openapi](api:endpoints.documented.openapi.package) and
-[endpoints.documented.delegate](api:endpoints.documented.delegate.package) packages).
```
done

Krever commented on documentation/manual/src/ornate/tutorial-openapi.md:
```diff
@@ -75,18 +75,16 @@ modifying it.
-interfaces available in this package allow us to supply documentation information such as
-human readable descriptions of request and response entities.
+[endpoints.algebra](api:endpoints.algebra.package). We use the same algebra for defining documentation
+as we would use for definig protocols for client/server imterpreatation.
```
done

julienrf commented on algebras/algebra/src/main/scala/endpoints/Semigroupal.scala:
```diff
@@ -0,0 +1,15 @@
+  implicit class SemigroupalSyntax[A, F[_]](val f: F[A])(implicit ev: Semigroupal[F]) {
+    def ++[B](other: F[B])(implicit tupler: Tupler[A, B]): F[tupler.Out] = ev.add(f,other)
+  }
+}
```
No, that’s enough, thanks!

julienrf commented on documentation/manual/src/ornate/interfaces-and-interpreters.md:
```diff
@@ -45,3 +45,7 @@ navigate through the algebra interfaces hierarchy.
 |[endpoints-scalaj-client](https://index.scala-lang.org/julienrf/endpoints/endpoints-scalaj-client)|JVM client backed by [scalaj-http](https://github.com/scalaj/scalaj-http)|
+|[endpoints-sttp-client](https://index.scala-lang.org/julienrf/endpoints/endpoints-sttp-client)|JVM client backed by [sttp](https://github.com/softwaremill/sttp)|
+|[endpoints-openapi](https://index.scala-lang.org/julienrf/endpoints/endpoints-openapi)|Interpreter that creates OpenAPI json from endpoints description |
+|[endpoints-openapi-json-schema-generic](https://index.scala-lang.org/julienrf/endpoints/endpoints-openapi-json-schema-generic)|Interpreter that allows generic derivation of json-schema|
```
I think the new name should be `endpoints-jsonschema-generic`. Wdyt?

julienrf commented on algebras/algebra/src/main/scala/endpoints/Semigroupal.scala:
```diff
@@ -0,0 +1,15 @@
+
+trait Semigroupal[F[_]] {
+
+  def add[A, B](fa: F[A], fb: F[B])(implicit tupler: Tupler[A, B]): F[tupler.Out]
```
OK :)

Krever commented on documentation/examples/documented/src/main/scala/counter/Counter.scala:
```diff
@@ -84,24 +89,19 @@ object CounterDocumentation
 //#openapi
 
 // Implementation of the HTTP API and its business logic
-//#delegation
```
Good catch, I have missed it originally but should be fixed now.

Krever commented:
Thanks a lot for the review effort. I promise not to ever again do such a huge change in a single PR :)

ches commented:
NB: it looks like the docs site needs to be regenerated since this merge.
